Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+            }
+        }
+
+        mutating func timeoutWaiter(_ requestID: RequestID) -> Action {
+            // 1. check waiters in starting connections
+            let connectionIndex = self.connections.firstIndex(where: {
+                switch $0.state {
+                case .starting(let waiter):
+                    return waiter?.requestID == requestID
+                case .active, .failed, .closed:
+                    return false
+                }
+            })
+
+            if let connectionIndex = connectionIndex {
+                var connectionState = self.connections[connectionIndex]
+                var requestAction: StateMachine.RequestAction = .none
+                if let waiter = connectionState.removeStartWaiter() {
+                    requestAction = .failRequest(waiter.request, HTTPClientError.connectTimeout, cancelWaiter: nil)
+                }
+                self.connections[connectionIndex] = connectionState
+
+                return .init(requestAction, .none)
+            }
+
+            // 2. check waiters in queue
+            let waiterIndex = self.waiters.firstIndex(where: { $0.requestID == requestID })
+            if let waiterIndex = waiterIndex {
+                // TBD: This is slow. Do we maybe want something more sophisticated here?
+                let waiter = self.waiters.remove(at: waiterIndex)
+                return .init(
+                    .failRequest(waiter.request, HTTPClientError.getConnectionFromPoolTimeout, cancelWaiter: nil),
+                    .none
+                )
+            }
+
+            // 3. we reach this point, because the waiter may already have been scheduled. The waiter
+            //    was not cancelled because of a race condition
+            return .init(.none, .none)
+        }
+
+        mutating func cancelWaiter(_ requestID: RequestID) -> Action {
+            // 1. check waiters in starting connections
+            let connectionIndex = self.connections.firstIndex(where: {
+                switch $0.state {
+                case .starting(let waiter):
+                    return waiter?.requestID == requestID
+                case .active, .failed, .closed:
+                    return false
+                }
+            })
+
+            if let connectionIndex = connectionIndex {
+                var connectionState = self.connections[connectionIndex]
+                var requestAction: StateMachine.RequestAction = .none
+                if let waiter = connectionState.removeStartWaiter() {
+                    requestAction = .failRequest(waiter.request, HTTPClientError.cancelled, cancelWaiter: waiter.requestID)
+                }
+                self.connections[connectionIndex] = connectionState
+
+                return .init(requestAction, .none)
+            }
+
+            // 2. check waiters in queue
+            let waiterIndex = self.waiters.firstIndex(where: { $0.requestID == requestID })
```

Comment:
Cancel and timeout here appear to be identical except for error type. Refactor to be common.

Reply:
The action is also different. In the timeout case, we need to cancel the request. If the request was cancelled, we don't need to fail the request anymore. I personally would like to keep them separate. But not a hill I want to die on.